feat: csp report service

mohdnr opened the pull request:
New security service to allow CDS apps to submit content security policy failures. This will be done by adding the `report-uri` csp directive as follow:
- `report-uri https://csp-report-to.security.cdssandbox.xyz/report;`

Reports will be available on the new csp violation app hosted behind the security tools SSO proxy: https://csp-reports.security.cdssandbox.xyz/

![image](https://user-images.githubusercontent.com/85885638/182008863-62d1a3ec-139e-4d99-aa4d-5af072d3e9fa.png)


Note: Majority of changes applied locally so plan should have limited outputs.

---

`report-to` support will be added at a later date. `report-to` format is as follow:
- `report-to {"group":"csp-endpoint","max_age":10886400,"endpoints":[{"url":"https://csp-report-to.security.cdssandbox.xyz/report"}]}";`

Commit:
feat: csp report service

Commit:
feat: tighten post rules for proxy

Commit:
chore: run tests for new app

Commit:
chore: missing makefile targets

patheard commented on .github/workflows/ci_code.yml:
```diff
@@ -22,6 +22,7 @@ jobs:
         with:
           filters: |
             images/cloud_asset_inventory/sentinel_neo4j_forwarder/src: 'images/cloud_asset_inventory/sentinel_neo4j_forwarder/src/**'
+            images/csp_violation_report_service/app: 'images/csp_violation_report_service/app/**'
```
Do we need a change to `pull_request/paths` trigger section as well?

Commit:
chore: missing workflow path

patheard commented on images/csp_violation_report_service/app/config/application.py:
```diff
@@ -0,0 +1,16 @@
+from masonite.environment import env
+
+
+KEY = env("APP_KEY", "-RkDOqXojJIlsF_I8wWiUq_KRZ0PtGWTOZ676u5HtLg=")
```
What do you think of the idea of not passing a default `APP_KEY` here and then failing if one doesn't exist?

The scenario I'm thinking of is us accidentally using this `APP_KEY` live because we forget to specify one in the `.env`.

patheard commented on images/csp_violation_report_service/app/config/mail.py:
```diff
@@ -0,0 +1,24 @@
+    "terminal": {
+        "from": FROM_EMAIL,
+    },
+}
```
Since we're not using this, can Masonite start without it?  My worry is that it'll make someone in the future think we do have these env vars setup.

patheard commented on images/csp_violation_report_service/app/config/notification.py:
```diff
@@ -0,0 +1,20 @@
+        "key": env("VONAGE_KEY", ""),
+        "secret": env("VONAGE_SECRET", ""),
+        "sms_from": env("VONAGE_SMS_FROM", "+33000000000"),
+    },
```
Same comment about removing if possible.

patheard commented on images/csp_violation_report_service/app/databases/seeds/database_seeder.py:
```diff
@@ -0,0 +1,10 @@
+class DatabaseSeeder(Seeder):
+    def run(self):
+        """Run the database seeds."""
+        # self.call(UserTableSeeder)
```
I'm guessing this is only something we'd need for local dev?

mohdnr replied:
preview app as well but i don't have plans for that at the moment even tho masonite makes it super simple by supporting sqlite databases.

patheard commented on images/csp_violation_report_service/app/makefile:
```diff
@@ -0,0 +1,22 @@
+install:
```
Can you add some `.PHONY` targets for these:
```makefile
.PHONY: install install-dev lint # etc
```

patheard commented on images/csp_violation_report_service/app/requirements.txt:
```diff
@@ -0,0 +1,7 @@
+aws-lambda-powertools==1.26.0
+masonite>=4.0,<5.0
+masonite-orm>=2.0,<3.0
```
Could we pin both of these to a specific version to cut down on any weirdness during version bumps and the build process?

patheard commented on images/csp_violation_report_service/app/templates/base.html:
```diff
@@ -0,0 +1,66 @@
+</head>
+
+<body>
+  <nav aria-label="Skip to main content">
```
French translation svp:
```
Passer au contenu principal
```

patheard commented on images/csp_violation_report_service/app/templates/base.html:
```diff
@@ -0,0 +1,66 @@
+      <div data-testid="fip" class="gc-fip">
+        <div class="canada-flag"><a href="http://canada.ca/{{t('language_code')}}.html"
+            aria-label="{{t('gc_home')}}"><img src="/sig-blk-{{t('language_code')}}.svg"
+              alt="Symbol of the Government of Canada"></a>
```
French translation:
```
Symbole du gouvernement du Canada
```

patheard commented on images/csp_violation_report_service/app/templates/base.html:
```diff
@@ -0,0 +1,66 @@
+        </div>
+        <div class="gc-language-toggle">
+          <label for="lang-switcher-button">
+            Language Selection
```
```
Sélection de la langue
```

patheard commented on images/csp_violation_report_service/app/templates/base.html:
```diff
@@ -0,0 +1,66 @@
+    </main>
+    <footer class="gc-footer" data-testid="footer">
+      <div class="gc-footer-container items-start">
+        <div><img alt="Symbol of the Government of Canada" src="/wmms-blk.svg"></div>
```
Same comment for French translation

patheard commented on images/csp_violation_report_service/app/templates/errors/403.html:
```diff
@@ -0,0 +1,20 @@
+  <meta charset="UTF-8">
+  <meta http-equiv="X-UA-Compatible" content="IE=edge">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <title>Permission Denied</title>
```
I'll create a bug to look at creating lang specific error pages.

patheard commented on images/csp_violation_report_service/app/templates/home.html:
```diff
@@ -0,0 +1,70 @@
+  </table>
+  <nav class="relative z-0 inline-flex rounded-md shadow-sm -space-x-px" aria-label="Pagination">
+    <a href="/?page=1" class="relative inline-flex items-center px-2 py-2 rounded-l-md border border-gray-300 bg-white text-sm font-medium text-gray-500 hover:bg-gray-50">
+      <span class="sr-only">Beginning</span>
```
Translation of `Début` could work here.

patheard commented on images/csp_violation_report_service/app/tests/unit/controllers/test_ReportController.py:
```diff
@@ -0,0 +1,151 @@
+            "referrer": "https://www.google.com/",
+            "violated-directive": "default-src self",
+            "original-policy": "default-src self; report-uri /csp-hotline.php",
+            "blocked-uri": "http://evilhackerscripts.com",
```
I mean the fact that this is http only is just a sad comment on the state of evil hackers in general.

patheard commented on terragrunt/aws/csp_violation_report_service/lambda.tf:
```diff
@@ -0,0 +1,56 @@
+    DB_HOST                 = aws_ssm_parameter.db_host.value
+    DB_USERNAME             = aws_ssm_parameter.db_username.value
+    DB_DATABASE             = aws_ssm_parameter.db_database.value
+    DB_PASSWORD             = aws_ssm_parameter.db_password.value
```
Could these be loaded on cold start directly from SSM rather than as `environment_variables`?  I'm permanently scared by Notify because of this...

patheard commented on images/csp_violation_report_service/app/bin/entry.sh:
```diff
@@ -0,0 +1,50 @@
+  done < $TMP_ENV_FILE
+}
+
+if [ ! -f "$ENV_PATH/.env" ]; then # Only setup envs once per lambda lifecycle
```
nit: could you change this comment so it references ECS task startup instead of lambda

patheard reviewed (changes requested):
Really nice stuff - Masonite looks like a nice framework.

patheard committed:
chore: add .PHONY targets

patheard committed:
fix: update i18n translations for base template

patheard committed:
chore: pin masonite dependency versions

patheard commented on terragrunt/aws/csp_violation_report_service/iam_lambda.tf:
```diff
@@ -0,0 +1,29 @@
+      module.purge_csp_reports_lambda.function_arn
+    ]
+  }
+}
```
Just noticed this - since this policy is only used by the Lambda function, does it need to the ability to invoke itself?  From what I could see in the code, the answer is no, but wanted to double check before removing.

mohdnr replied:
i think this is a copy paste fail from another project. Totally don't need this @patheard

patheard committed:
fix: load SSM parameters on lambda cold start

This prevents accidental leakage of secret values
by CLI and Terraform operations.

patheard committed:
fix: remove un-used IAM policy

patheard committed:
chore: add Docker SBOM generate for CSP scanner

Commit:
chore: remove unused items and default to no APP_KEY

Commit:
chore: ci setup .env

Commit:
chore: point to local

Commit:
chore: more ci fixes
